### DISCUSSION TITLE
Keeping connection open vs opening and closing connection when reading values periodically

### DISCUSSION THREAD

**Discussion by fcapallera** [FreeOpcUa/python-opcua#1421] [Q&A] [answered]
Hi,

I'm currently developing a program that reads data from a PLC through OPC periodically (every 5-10 seconds) and sends it to the cloud. The process is supposed to keep running all day.

I was wondering if it's better practice to open the OPC connection, read the values and close the connection once the program ends, or open the connection to read the value, close the connection and then open it again after 5-10 seconds when I need to read the value again.

Opening/closing connection once per session or opening/closing connection once per data read (5-10 seconds)?

Thanks!

**Comment by AndreasHeine** [comment, marked as answer]
keep the connection open! we currently cant reuse the session... so each connect is a new session in the plc and depending on the session timout it can lead to DoS (session exhaustion) of the OPC UA Server.

**Reply by fcapallera** [reply]
Thanks for the quick reply!

**Comment by oroulet** [comment]
@AndreasHeine do we have any issues with sessions staying alive on server? as far as I know we send a request to server to close session. So having sessions left on server should only happen if there is a crash or network issue

**Reply by AndreasHeine** [reply]
in some products ua servers had issues in the past but its not on our side! typically a session is not bound to a connection so it should be long living. connecting and disconnecting in an interval of 5-10s is not wise from my point of view, it would be ok if the sessiontimeout is smaller then the interval! but sessions should not have like less then 30s timeouts, from an security perspective it is also not wise to always connect and disconnect because you will send credentials in each create/activate session which will be decodeable (even with asymetric algos) with enougth recording of that handshake... but thats my POV!
